add task list output - resolves #30

SamVerschueren opened the pull request:
I implemented [listr](https://github.com/samverschueren/listr) in order to show the task list that will be executed upon publishing.

I tested it and it seems to work as expected. But might be a good idea for someone else to test this as well :).

--------------------
I converted `execa` to an `Observable` because it is not yet supported by `execa` itself. Is this good enough (for now)?  @jamestalmage

jamestalmage commented:
> I converted execa to an Observable because it is not yet supported by execa itself. Is this good enough (for now)?

I say yes! I don't want to wait for `execa` on this, and besides, it's probably better to see a few good use cases before locking in the `execa` API.

A couple thoughts:

- Use `stream-to-observable`. I have a PR open to switch it to use `any-observable`.

- Use `split` to ensure each data event is for a single line:

- Use `rxjs` Observables instead of `zen-observables`

- Use the [`merge` method from rxjs](http://reactivex.io/rxjs/class/es6/Observable.js~Observable.html#static-method-merge) to merge `stdout` and `stderr`

- Make sure you get the full `rxjs` (needed for Merge) by requiring: `any-observable/register/rxjs-all`.

  ```js
  const cp = execa(...);
  const observable = Observable.merge(
    streamToObservable(cp.stdout.pipe(split())),
    streamToObservalbe(cp.stderr.pipe(split()))
  );
  ```

sindresorhus commented:
> Use split to ensure each data event is for a single line:

Maybe `stream-to-observable` should have an option for this? Seems like a very common use-case, and many don't know about the `split` module. I actually can't think of a use-case where I wouldn't want it split on lines.

jamestalmage commented:
@SamVerschueren - `stream-to-observable@0.2.0` is out. It should be easier to use now.

-- 

> Maybe stream-to-observable should have an option for this?

Yeah sure. https://github.com/jamestalmage/stream-to-observable/issues/8

> I actually can't think of a use-case where I wouldn't want it split on lines.

Agreed, it's probably the most common case. Not all streams are binary text data though (i.e. Object streams), so I am not sure it's a good idea to apply by default.

jamestalmage commented:
I'm really very excited about this PR BTW.

SamVerschueren commented:
You're right about `Observable.merge` and `stream-to-observable`. Really nice solution!

SamVerschueren commented on index.js:
```diff
@@ -2,27 +2,61 @@
+	return Observable.merge(
+		streamToObservable(cp.stdout.pipe(split())),
+		streamToObservable(cp.stderr.pipe(split()))
+	).filter(Boolean);
```
Is this correct? I only want the non-empty lines.

SamVerschueren replied:
Will the observable reject if `cp` rejects? Not sure about this being the same as my previous attempt :).

jamestalmage replied:
> Is this correct? I only want the non-empty lines.

Seems reasonable to me.

> Will the observable reject if cp rejects?

See the [`await`](https://github.com/jamestalmage/stream-to-observable#await) option on `stream-to-observable`.

SamVerschueren replied:
> See the await option on stream-to-observable.

Cool! Added.

sindresorhus commented:
Such beautiful output!

sindresorhus commented:
I noticed that the spinner is a bit jittery during the `Cleanup` task. Any idea why?

sindresorhus commented:
When it's finished, should we replace all the task output with something like this?

```
✔ package-name 1.2.3 published
```

sindresorhus commented on index.js:
```diff
@@ -33,44 +67,54 @@ module.exports = (input, opts) => {
-	log('Done! 🎉');
+	tasks.add([
+		{
+			title: 'Update version',
```
> Bumping version

?

SamVerschueren replied:
👍

SamVerschueren commented:
> I noticed that the spinner is a bit jittery during the Cleanup task (when the node_modules folder is large). Any idea why? Maybe something blocking.

Noticed that as well. It must be something blocking in the `del` module. When I use `rm -r node_modules`, it runs smoothly.

```js
	if (runCleanup) {
		tasks.add([
			{
				title: 'Cleanup',
				task: () => exec('rm', ['-r', 'node_modules'])
			},
			{
				title: 'Installing dependencies',
				task: () => exec('npm', ['install'])
			}
		]);
	}
```

SamVerschueren commented:
Added a success message

```
 ✔ Git
 ✔ Cleanup
 ✔ Installing dependencies
 ✔ Running tests

 mem-map 1.0.1 published
```

Added the checkmark before the success message but I think it's too much :p. Everything green indicates success and that message is informative. I looked nicer without checkmark (for me).

SamVerschueren commented:
I also tried the cleanup task with `rimraf` directly and that package seems to cause the jittery spinner.

SamVerschueren commented on index.js:
```diff
@@ -1,28 +1,65 @@
+	return Observable.merge(
+		streamToObservable(cp.stdout.pipe(split()), {await: cp}),
+		streamToObservable(cp.stderr.pipe(split()), {await: cp})
+	).filter(Boolean);
```
Not sure if necessary but should we add trimming here to make sure the offset is gone. Did not saw an example of that but it seems quite obvious that it's possible.

```js
return Observable.merge(
	streamToObservable(cp.stdout.pipe(split()), {await: cp}),
	streamToObservable(cp.stderr.pipe(split()), {await: cp})
).map(x => x.trim()).filter(Boolean);
```

jamestalmage replied:
Should trimming the event strings maybe be a part of `listr`?

SamVerschueren replied:
You're right, that would be better!

SamVerschueren replied:
Added it in `listr` and bumped version.

jamestalmage commented:
Just used it to deploy. It's awesome. :shipit: !!

sindresorhus commented:
I've been using this for a week now and it's been working perfectly.

sindresorhus commented:
Thank you so much for doing this @SamVerschueren. You made this module so much better! :)

![68747470733a2f2f6d656469612e67697068792e636f6d2f6d656469612f695054546a4574313969676e652f67697068792e676966](https://cloud.githubusercontent.com/assets/170270/16631967/1ca54f4c-43c2-11e6-88e4-eccd4e78ad4f.gif)

SamVerschueren commented:
Very excited this got released!! :D

jfmengels commented:
Just wanted to say this is really great. Thanks @SamVerschueren!
![season-4-the-simpsons-4x2-l2JdVRfJozpjq70SA](https://camo.githubusercontent.com/41fbfd2e06c5c4ae82133cf03c2190e2edbe6aa9/68747470733a2f2f6d65646961322e67697068792e636f6d2f6d656469612f6c324a645652664a6f7a706a71373053412f67697068792e676966>)

SamVerschueren commented:
Thanks @jfmengels ! :)
